packages/table/src/quadrants/tableQuadrant.tsx

```diff
@@ -0,0 +1,161 @@
+
+    /**
+     * A callback that renders the table menu (the rectangle in the top-left corner).
+     */
+    renderMenu?: () => JSX.Element;
+
+    /**
+     * A callback that renders either all of or just the frozen section of the column header.
+     */
+    renderColumnHeader?: (showFrozenColumnsOnly?: boolean) => JSX.Element;
+
+    /**
+     * A callback that renders either all of or just the frozen section of the row header.
+     */
+    renderRowHeader?: (showFrozenRowsOnly?: boolean) => JSX.Element;
+
+    /**
+     * A callback that renders either all of or just frozen sections of the table body.
+     */
+    renderBody?: (showFrozenRowsOnly?: boolean, showFrozenColumnsOnly?: boolean) => JSX.Element;
+
+    /**
+     * A callback that receives a `ref` to the quadrant's scroll-container element.
+     */
+    scrollContainerRef: React.Ref<HTMLElement>;
+
+    /**
+     * CSS styles to apply to the quadrant's outermost element.
+     */
+    style?: React.CSSProperties;
+}
+
+export class TableQuadrant extends AbstractComponent<ITableQuadrantProps, {}> {
+    public render() {
+        const { quadrantType } = this.props;
+
+        const showFrozenRowsOnly = quadrantType === QuadrantType.TOP || quadrantType === QuadrantType.TOP_LEFT;
+        const showFrozenColumnsOnly = quadrantType === QuadrantType.LEFT || quadrantType === QuadrantType.TOP_LEFT;
+
+        const className = classNames(Classes.TABLE_QUADRANT, this.getQuadrantCssClass(), this.props.className);
+
+        return (
+            <div className={className} style={this.props.style} ref={this.props.quadrantRef}>
+                <div
+                    className={Classes.TABLE_QUADRANT_SCROLL_CONTAINER}
+                    ref={this.props.scrollContainerRef}
+                    onScroll={this.props.onScroll}
+                    onWheel={this.props.onWheel}
+                >
+                    <div className={Classes.TABLE_TOP_CONTAINER}>
+                        {this.props.renderMenu()}
+                        {this.props.renderColumnHeader(showFrozenColumnsOnly)}
+                    </div>
+                    <div className={Classes.TABLE_BOTTOM_CONTAINER}>
+                        {this.props.renderRowHeader(showFrozenRowsOnly)}
+                        {/* TODO: is it okay to put `position: relative` on every quadrant's body wrapper? */}
+                        <div
+                            className={Classes.TABLE_QUADRANT_BODY_CONTAINER}
+                            ref={this.props.bodyRef}
+                        >
+                            {this.props.renderBody(showFrozenRowsOnly, showFrozenColumnsOnly)}
+                        </div>
+                    </div>
+                </div>
+            </div>
+        );
+    }
+
+    protected validateProps(nextProps: ITableQuadrantProps) {
+        const { quadrantType } = nextProps;
+        if (nextProps.onScroll != null && quadrantType !== QuadrantType.MAIN) {
```

Comment:
Interesting. Would this require two new components too?

`<MainQuadrant {...mainQuadrantProps} />`
`<NonMainQuadrant {...topQuadrantProps} />`
`<NonMainQuadrant {...leftQuadrantProps} />`
`<NonMainQuadrant {...topLeftQuadrantProps} />`